### PR TITLE
Docs: Generate note about extensions module in message documentation

### LINKED ISSUE #6051: Add information about extension enabling in messages documentation
### Current problem

Current message documentation does not provide information whether message (e.g. bad-builtin) is extension.

### Desired solution

Message documentation should provide note that message is part of extension and provide link to the extension.

### Additional context

See https://github.com/PyCQA/pylint/pull/6022 for details

### PR DESCRIPTION
**Pull request by matusvalo**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
|     | :bug: Bug fix          |
| ✓   | :sparkles: New feature |
|     | :hammer: Refactoring   |
| ✓   | :scroll: Docs          |

## Description

Closes #6051

### CODE REVIEW

File: doc/exts/pylint_messages.py [hunk 1]
```diff
@@ -179,10 +186,17 @@ def _write_message_page(messages_dict: MessagesDict) -> None:
...
 {message.related_links}
-
-Created by ``{message.checker}`` checker
 """
                 )
+                if message.checker_module.startswith("pylint.extensions."):
+                    stream.write(
+                        f"""
+.. note::
+  This check requires ``{message.checker_module}`` plugin to be loaded. See: :ref:`{message.checker_module}`
```
Comment (DanielNoord) [inline doc/exts/pylint_messages.py:195]:
```suggestion
  This check requires the ``{message.checker_module}`` plugin to be loaded. See: :ref:`{message.checker_module}`
```

File: doc/exts/pylint_messages.py [hunk 2]
```diff
@@ -179,10 +186,17 @@ def _write_message_page(messages_dict: MessagesDict) -> None:
...
+
+"""
+                    )
+                stream.write(f"""Created by ``{message.checker}`` checker""")
```
Comment (DanielNoord) [inline doc/exts/pylint_messages.py:199]: While we're at it. Originally I was thinking of adding a link to the code on Github here. Something like `https://github.com/PyCQA/pylint/blob/main/pylint/checkers/CHECKER_NAME` should work? The only thing we should be careful about is checkers in subdirectories.

Is this something you would want to explore in this PR as well?

Comment (matusvalo) [reply doc/exts/pylint_messages.py:199]: I will check that.

**Review by Pierre-Sassoulas — APPROVED** [review]
👌

File: doc/exts/pylint_messages.py (hunk 2 above)
Comment (matusvalo) [reply doc/exts/pylint_messages.py]: Implemented

**Comment by matusvalo** [comment]
I have implemented the link to to checker file in github plus I have improved note.

File: doc/exts/pylint_messages.py [hunk 3]
```diff
@@ -29,6 +31,8 @@
 
 MSG_TYPES_DOC = {k: v if v != "info" else "information" for k, v in MSG_TYPES.items()}
 
+BASE_PATH = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
```
Comment (matusvalo) [inline doc/exts/pylint_messages.py:34]: I think we can use PYLINT_BASE_PATH instead.

Comment (Pierre-Sassoulas) [reply doc/exts/pylint_messages.py:34]:
```suggestion
PYLINT_BASE_PATH = Path(_file__).parent.parent.resolve()
```

File: doc/exts/pylint_messages.py [hunk 4]
```diff
@@ -39,6 +42,8 @@ class MessageData(NamedTuple):
     bad_code: str
     details: str
     related_links: str
+    checker_module_name: str
```
Comment (DanielNoord) [inline doc/exts/pylint_messages.py:45]: Is there a difference between this and `checker`, defined on L37?

File: doc/exts/pylint_messages.py [hunk 5]
```diff
@@ -126,12 +131,21 @@ def _get_all_messages(
...
+    )
+    for checker, message in checker_message_mapping:
         message_data_path = (
             PYLINT_MESSAGES_DATA_PATH / message.symbol[0] / message.symbol
         )
         good_code, bad_code, details, related = _get_message_data(message_data_path)
 
+        checker_module = getmodule(checker)
+
+        if not checker_module or not checker_module.__file__:
```
Comment (DanielNoord) [inline doc/exts/pylint_messages.py:146]: For which checkers does this fail?

File: doc/exts/pylint_messages.py [hunk 6]
```diff
@@ -126,12 +131,21 @@ def _get_all_messages(
...
+            sys.exit(f"Cannot find module for hecker {checker}")
```
Comment (DanielNoord) [inline doc/exts/pylint_messages.py:147]:
```suggestion
            sys.exit(f"Cannot find module for checker {checker}")
```

File: doc/exts/pylint_messages.py [hunk 7]
```diff
@@ -179,9 +198,19 @@ def _write_message_page(messages_dict: MessagesDict) -> None:
 {message.bad_code}
 {message.details}
 {message.related_links}
+"""
+                )
+                if message.checker_module_name.startswith("pylint.extensions."):
+                    stream.write(
+                        f"""
+.. note::
+  This message is emited by the checker which requires the ``{message.checker_module_name}`` plugin to be loaded. See: :ref:`{message.checker_module_name}`
```
Comment (DanielNoord) [inline doc/exts/pylint_messages.py:207]:
```suggestion
  This message is emitted by an optional checker which requires the ``{message.checker_module_name}`` plugin to be loaded. See: :ref:`{message.checker_module_name}`
```

**Review by DanielNoord — COMMENTED** [review]
Thanks for adding the second part as well @matusvalo 😄

File: doc/exts/pylint_messages.py (hunk 5 above)
Comment (matusvalo) [reply doc/exts/pylint_messages.py:146]: It should not fall. It was introduced to make mypy happy.

Comment (DanielNoord) [reply doc/exts/pylint_messages.py:146]: Ah, can you use an assert then? That way we will also fail the build process if for some reason this does fail in the future.

File: doc/exts/pylint_messages.py (hunk 4 above)
Comment (matusvalo) [reply doc/exts/pylint_messages.py:45]: checker is the name of checker. Checker_module_name is full name of checker module e.g. pylint.extensions.bad_builtin

Comment (DanielNoord) [reply doc/exts/pylint_messages.py:45]: Ah of course. Makes sense!

**Comment by matusvalo** [comment]
Assert added, typo fixed.

File: doc/exts/pylint_messages.py [hunk 8]
```diff
@@ -179,9 +198,19 @@ def _write_message_page(messages_dict: MessagesDict) -> None:
 {message.bad_code}
 {message.details}
 {message.related_links}
+"""
+                )
+                if message.checker_module_name.startswith("pylint.extensions."):
+                    stream.write(
+                        f"""
+.. note::
+  This message is emitted by an optional checker which requires the ``{message.checker_module_name}`` plugin to be loaded. See: :ref:`{message.checker_module_name}`
```
Comment (DanielNoord) [inline doc/exts/pylint_messages.py:207]:
```suggestion
  This message is emitted by an optional checker which requires the ``{message.checker_module_name}`` plugin to be loaded. See: :ref:`{message.checker_module_name}`.
```

File: doc/exts/pylint_messages.py [hunk 9]
```diff
@@ -179,9 +198,19 @@ def _write_message_page(messages_dict: MessagesDict) -> None:
...
 
-Created by ``{message.checker}`` checker
 """
+                    )
+                checker_url = f"https://github.com/PyCQA/pylint/blob/main/{checker_module_rel_path}"
+                stream.write(
+                    f"Created by `{message.checker} <{checker_url}>`__ checker"
```
Comment (DanielNoord) [inline doc/exts/pylint_messages.py:213]:
```suggestion
                    f"Created by the `{message.checker} <{checker_url}>`__ checker."
```

**Review by DanielNoord — APPROVED** [review]
Two final comments. Thanks @matusvalo, this is most definitely an improvement!

**Pull request merged**